### PR TITLE
fix: Require package names to be non-empty

### LINKED ISSUE #2287: `CrateName` does not reject empty strings
### Aim

`nargo compile` shouldn't produce an empty target dir when the artifact (target.json) is placed in the project root.

### Expected Behavior

see above.

### Bug

no error messages.

### To Reproduce

1. run `nargo compile`. see output described above

### Installation Method

None

### Nargo Version

nargo 0.9.0 (git version hash: 3d1b2522c8d9a0acebff102f9f877c44178c5db5, is dirty: false)

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves #2287 

## Summary\*

This PR requires that `CrateNames` must have a non-empty string to avoid the issue described in #2287 

We also weren't validating these on package creation, so I've added a custom value parser to clap in order to error on invalid user input as soon as possible so we don't get `nargo new` half-creating the new package.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

BEGIN_COMMIT_OVERRIDE
fix: Require package names to be non-empty (#2293)
feat: Perform input validation on user's package names (#2293)
END_COMMIT_OVERRIDE

### CODE REVIEW

**Review by phated — CHANGES_REQUESTED** [review]
This should just be implemented in the `FromStr` trait on CrateName

File: crates/nargo_cli/src/cli/init_cmd.rs [hunk 1]
```diff
@@ -6,13 +6,15 @@ use acvm::Backend;
...
+use noirc_frontend::graph::CrateName;
 use std::path::PathBuf;
+use std::str::FromStr;
 
 /// Create a Noir project in the current directory.
 #[derive(Debug, Clone, Args)]
 pub(crate) struct InitCommand {
     /// Name of the package [default: current directory name]
-    #[clap(long)]
+    #[clap(long, value_parser = CrateName::from_str)]
```
Comment (phated) [inline crates/nargo_cli/src/cli/init_cmd.rs:17]:
```suggestion
    #[clap(long)]
```
Clap does this by default if you change `Option<String>` to `Option<CrateName>`

File: crates/nargo_cli/src/cli/new_cmd.rs [hunk 2]
```diff
@@ -13,7 +14,7 @@ pub(crate) struct NewCommand {
     path: PathBuf,
 
     /// Name of the package [default: package directory name]
-    #[clap(long)]
+    #[clap(long, value_parser = CrateName::from_str)]
```
Comment (phated) [inline crates/nargo_cli/src/cli/new_cmd.rs:17]: See my other comment

File: crates/noirc_frontend/src/graph/mod.rs [hunk 3]
```diff
@@ -54,11 +60,28 @@ impl FromStr for CrateName {
...
 
     fn from_str(name: &str) -> Result<Self, Self::Err> {
-        let is_invalid = name.chars().any(|n| CHARACTER_BLACK_LIST.contains(&n));
-        if is_invalid {
-            Err(name.into())
-        } else {
+        if Self::is_valid_name(name) {
             Ok(Self(SmolStr::new(name)))
+        } else {
+            Err("Package names must be non-empty and cannot contain hyphens".into())
```
Comment (phated) [inline crates/noirc_frontend/src/graph/mod.rs:66]: I like bringing in this better error message, but should we also inject the package name? Something like:
```suggestion
            Err(format!("Package names must be non-empty and cannot contain hyphens, but received {name}").into())
```

File: crates/noirc_frontend/src/graph/mod.rs [hunk 4]
```diff
@@ -29,6 +29,12 @@ impl CrateId {
 #[derive(Debug, Clone, PartialEq, Eq, Hash, Ord, PartialOrd)]
 pub struct CrateName(SmolStr);
 
+impl CrateName {
+    pub fn is_valid_name(name: &str) -> bool {
```
Comment (phated) [inline crates/noirc_frontend/src/graph/mod.rs:33]:
```suggestion
   fn is_valid_name(name: &str) -> bool {
```
Don't make it pub unless necessary.

**Review by phated — CHANGES_REQUESTED** [review]
A few things, but looking good

File: crates/noirc_frontend/src/graph/mod.rs (hunk 3 above)
Comment (TomAFrench) [reply crates/noirc_frontend/src/graph/mod.rs:66]: Clap displays the bad package name by default.

File: crates/noirc_frontend/src/graph/mod.rs (hunk 4 above)
Comment (TomAFrench) [reply crates/noirc_frontend/src/graph/mod.rs:33]: Ah, forgot to remove this in last commit.

File: crates/noirc_frontend/src/graph/mod.rs (hunk 3 above)
Comment (phated) [reply crates/noirc_frontend/src/graph/mod.rs:66]: When we resolve a workspace, we map the string error to a variant and it doesn't contain the name at all. I'm going to push some updates that make this stuff more clear to this branch

**Comment by phated** [comment]
Should we add some failing integration tests for bad package names in Nargo.toml?

**Comment by TomAFrench** [comment]
> Should we add some failing integration tests for bad package names in Nargo.toml?

Done

**Review by phated — APPROVED** [review]

**Pull request merged**